Purchases/SubscriberAttributes/RCSubscriberAttributesManager.m

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2020 Purchases. All rights reserved.
+//
+
+#import "RCSubscriberAttributesManager.h"
+#import "RCSubscriberAttribute.h"
+#import "RCSpecialSubscriberAttributes.h"
+#import "RCBackend.h"
+#import "RCDeviceCache.h"
+#import "NSError+RCExtensions.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+
+@interface RCSubscriberAttributesManager ()
+
+@property (nonatomic) RCDeviceCache *deviceCache;
+@property (nonatomic) RCBackend *backend;
+
+@end
+
+
+@implementation RCSubscriberAttributesManager
+
+#pragma MARK - Public methods
+
+- (instancetype)initWithBackend:(nullable RCBackend *)backend
+                    deviceCache:(nullable RCDeviceCache *)deviceCache {
+    if (self = [super init]) {
+        self.backend = backend;
+        self.deviceCache = deviceCache;
+    }
+    return self;
+}
+
+- (void)setAttributes:(NSDictionary<NSString *, NSString *> *)attributes appUserID:(NSString *)appUserID {
+    [attributes enumerateKeysAndObjectsUsingBlock:^(id key, id value, BOOL *stop) {
+        [self setAttributeWithKey:key value:value appUserID:appUserID];
+    }];
+}
+
+- (void)setEmail:(nullable NSString *)email appUserID:(NSString *)appUserID {
+    [self setAttributeWithKey:SPECIAL_ATTRIBUTE_EMAIL value:email appUserID:appUserID];
+}
+
+- (void)setPhoneNumber:(nullable NSString *)phoneNumber appUserID:(NSString *)appUserID {
+    [self setAttributeWithKey:SPECIAL_ATTRIBUTE_PHONE_NUMBER value:phoneNumber appUserID:appUserID];
+}
+
+- (void)setDisplayName:(nullable NSString *)displayName appUserID:(NSString *)appUserID {
+    [self setAttributeWithKey:SPECIAL_ATTRIBUTE_DISPLAY_NAME value:displayName appUserID:appUserID];
+}
+
+- (void)setPushToken:(nullable NSString *)pushToken appUserID:(NSString *)appUserID {
+    [self setAttributeWithKey:SPECIAL_ATTRIBUTE_PUSH_TOKEN value:pushToken appUserID:appUserID];
+}
+
+- (void)syncIfNeededWithAppUserID:(NSString *)appUserID completion:(void (^)(NSError *_Nullable error))completion {
+    if ([self numberOfUnsyncedAttributesForAppUserID:appUserID] > 0) {
+        [self syncAttributesWithCompletion:completion appUserID:appUserID];
+    } else {
+        completion(nil);
+    }
+}
+
+- (RCSubscriberAttributeDict)unsyncedAttributesByKeyForAppUserID:(NSString *)appUserID {
+    return [self.deviceCache unsyncedAttributesByKeyForAppUserID:appUserID];
+}
+
+#pragma MARK - Private methods
+
+- (void)setAttributeWithKey:(nullable NSString *)key value:(NSString *)value appUserID:(NSString *)appUserID {
+    [self storeAttributeLocallyIfNeededWithKey:key value:value appUserID:appUserID];
+}
+
+- (void)syncAttributesWithCompletion:(void (^)(NSError *_Nullable error))completion appUserID:(NSString *)appUserID {
+    RCSubscriberAttributeDict unsyncedAttributes = [self unsyncedAttributesByKeyForAppUserID:appUserID];
+
+    [self.backend postSubscriberAttributes:unsyncedAttributes appUserID:appUserID completion:^(NSError *error) {
+        BOOL didBackendReceiveValues = (error == nil || error.successfullySynced);
```

Comment:
I'm printing it out in part 4 https://github.com/RevenueCat/purchases-ios/pull/190/files#diff-e76b90347a23e1c2409f6484538432f0R77. 
I'm also going to add the userInfo dict since it'll be useful, but I figured that would be enough, right?

Reply:
added to part 4: bae87f05c74842dfe93f6da8c1fcba80071d4912